### PR TITLE
Add shuffle data paramter to data_split

### PR DESCRIPTION
**Pull request by jeffin143**
In order to handle - https://github.com/mlpack/models/pull/60#discussion_r392193434

### CODE REVIEW

**Comment by kartikdutt18** [comment]
Thanks @jeffin143, I was about to start working on this for a new data loader. I think I had the same conversation with rcurtin a couple of days back. Thanks a lot.

File: src/mlpack/core/data/split_data.hpp [hunk 1]
```diff
@@ -61,9 +63,13 @@ void Split(const arma::Mat<T>& input,
...
-  const arma::Col<size_t> order =
-      arma::shuffle(arma::linspace<arma::Col<size_t>>(0, input.n_cols - 1,
-                                                      input.n_cols));
+  arma::Col<size_t> order;
+  if (shuffleData)
+    order = arma::shuffle(arma::linspace<arma::Col<size_t>>(0,
+        input.n_cols - 1, input.n_cols));
+  else
+    order = arma::linspace<arma::Col<size_t>>(0, input.n_cols - 1,
+        input.n_cols);
```
Comment (zoq) [inline src/mlpack/core/data/split_data.hpp]:
```suggestion
  if (shuffleData)
  {
    order = arma::shuffle(arma::linspace<arma::Col<size_t>>(0,
        input.n_cols - 1, input.n_cols));
  }
  else
  {
    order = arma::linspace<arma::Col<size_t>>(0, input.n_cols - 1,
        input.n_cols);
  }  
```

I think that is easier to read.

File: src/mlpack/core/data/split_data.hpp [hunk 2]
```diff
@@ -98,21 +104,27 @@ void Split(const arma::Mat<T>& input,
  * @param trainData Matrix to store training data into.
  * @param testData Matrix to store test data into.
  * @param testRatio Percentage of dataset to use for test set (between 0 and 1).
+ * @param shuffleData True(Default) if you want to shuffle data.
```
Comment (zoq) [inline src/mlpack/core/data/split_data.hpp]: What about: "If true, the sample order is shuffled; otherwise, each sample is visited in linear order. (Default true)."

File: src/mlpack/core/data/split_data.hpp [hunk 3]
```diff
@@ -98,21 +104,27 @@ void Split(const arma::Mat<T>& input,
...
-  const arma::Col<size_t> order =
-      arma::shuffle(arma::linspace<arma::Col<size_t>>(0, input.n_cols -1,
-                                                      input.n_cols));
+  arma::Col<size_t> order;
+  if (shuffleData)
+    order = arma::shuffle(arma::linspace<arma::Col<size_t>>(0,
+        input.n_cols -1, input.n_cols));
+  else
+    order = arma::linspace<arma::Col<size_t>>(0, input.n_cols -1,
+        input.n_cols);
```
Comment (zoq) [inline src/mlpack/core/data/split_data.hpp]:
```suggestion
  if (shuffleData)
  {
    order = arma::shuffle(arma::linspace<arma::Col<size_t>>(0,
        input.n_cols -1, input.n_cols));
  }
  else
  {
    order = arma::linspace<arma::Col<size_t>>(0, input.n_cols -1,
        input.n_cols);
  }
```

File: src/mlpack/tests/split_data_test.cpp [hunk 4]
```diff
@@ -108,6 +108,21 @@ BOOST_AUTO_TEST_CASE(SplitDataResultMat)
   CheckMatEqual(input, concat);
 }
 
+BOOST_AUTO_TEST_CASE(SplitDataResultMat)
+{
+  mat input(2, 10);
+  size_t count = 0; // count for putting unique sequential values
```
Comment (zoq) [inline src/mlpack/tests/split_data_test.cpp]: Hm, should this be "counter for unique sequential values"?

File: src/mlpack/tests/split_data_test.cpp [hunk 5]
```diff
@@ -108,6 +108,21 @@ BOOST_AUTO_TEST_CASE(SplitDataResultMat)
...
+  input.imbue([&count] () { return ++count; });
+
+  const auto value = Split(input, 0.2, false);
+  BOOST_REQUIRE_EQUAL(std::get<0>(value).n_cols, 8); // train data
```
Comment (zoq) [inline src/mlpack/tests/split_data_test.cpp]:
```suggestion
  BOOST_REQUIRE_EQUAL(std::get<0>(value).n_cols, 8); // Train data.
```

File: src/mlpack/tests/split_data_test.cpp [hunk 6]
```diff
@@ -108,6 +108,21 @@ BOOST_AUTO_TEST_CASE(SplitDataResultMat)
...
+  BOOST_REQUIRE_EQUAL(std::get<1>(value).n_cols, 2); // test data
```
Comment (zoq) [inline src/mlpack/tests/split_data_test.cpp]:
```suggestion
  BOOST_REQUIRE_EQUAL(std::get<1>(value).n_cols, 2); // Test data.
```

**Comment by ojhalakshya** [comment]
Hey @jeffin143 
I think by rebasing the branch will restart the tests, now the build tests are also working.
Thanks.

File: src/mlpack/tests/split_data_test.cpp (hunk 6 above)
Comment (jeffin143) [reply src/mlpack/tests/split_data_test.cpp]: Done

File: src/mlpack/tests/split_data_test.cpp (hunk 5 above)
Comment (jeffin143) [reply src/mlpack/tests/split_data_test.cpp]: Done

File: src/mlpack/tests/split_data_test.cpp (hunk 4 above)
Comment (jeffin143) [reply src/mlpack/tests/split_data_test.cpp]: Done

File: src/mlpack/core/data/split_data.hpp (hunk 3 above)
Comment (jeffin143) [reply src/mlpack/core/data/split_data.hpp]: Done

File: src/mlpack/core/data/split_data.hpp (hunk 2 above)
Comment (jeffin143) [reply src/mlpack/core/data/split_data.hpp]: Done

File: src/mlpack/core/data/split_data.hpp (hunk 1 above)
Comment (jeffin143) [reply src/mlpack/core/data/split_data.hpp]: Done

File: src/mlpack/tests/split_data_test.cpp [hunk 7]
```diff
@@ -94,20 +94,35 @@ void CheckDuplication(const Row<size_t>& trainLabels,
     BOOST_REQUIRE_EQUAL(counts[i], 1);
 }
 
-BOOST_AUTO_TEST_CASE(SplitDataResultMat)
+BOOST_AUTO_TEST_CASE(SplitShuffleDataResultMat)
 {
   mat input(2, 10);
-  size_t count = 0; // count for putting unique sequential values
+  size_t count = 0; // counter for unique sequential values
```
Comment (zoq) [inline src/mlpack/tests/split_data_test.cpp]:
```suggestion
  size_t count = 0; // Counter for unique sequential values.
```

To be consistent with the rest of the codebase.

File: src/mlpack/tests/split_data_test.cpp [hunk 8]
```diff
@@ -94,20 +94,35 @@ void CheckDuplication(const Row<size_t>& trainLabels,
...
   input.imbue([&count] () { return ++count; });
 
   const auto value = Split(input, 0.2);
-  BOOST_REQUIRE_EQUAL(std::get<0>(value).n_cols, 8); // train data
-  BOOST_REQUIRE_EQUAL(std::get<1>(value).n_cols, 2); // test data
+  BOOST_REQUIRE_EQUAL(std::get<0>(value).n_cols, 8); // Train data
```
Comment (zoq) [inline src/mlpack/tests/split_data_test.cpp]:
```suggestion
  BOOST_REQUIRE_EQUAL(std::get<0>(value).n_cols, 8); // Train data.
```

Pedantic style issue.

File: src/mlpack/tests/split_data_test.cpp [hunk 9]
```diff
@@ -94,20 +94,35 @@ void CheckDuplication(const Row<size_t>& trainLabels,
...
+  BOOST_REQUIRE_EQUAL(std::get<1>(value).n_cols, 2); // Test data
```
Comment (zoq) [inline src/mlpack/tests/split_data_test.cpp]:
```suggestion
  BOOST_REQUIRE_EQUAL(std::get<1>(value).n_cols, 2); // Test data.
```

Another pedantic style issue.

File: src/mlpack/tests/split_data_test.cpp [hunk 10]
```diff
@@ -94,20 +94,35 @@ void CheckDuplication(const Row<size_t>& trainLabels,
...
 
   mat concat = arma::join_rows(std::get<0>(value), std::get<1>(value));
   CheckMatEqual(input, concat);
 }
 
+BOOST_AUTO_TEST_CASE(SplitDataResultMat)
+{
+  mat input(2, 10);
+  size_t count = 0; // counter for unique sequential values
```
Comment (zoq) [inline src/mlpack/tests/split_data_test.cpp]:
```suggestion
  size_t count = 0; // Counter for unique sequential values.
```

Pedantic style issue number three :)

File: src/mlpack/tests/split_data_test.cpp [hunk 11]
```diff
@@ -94,20 +94,35 @@ void CheckDuplication(const Row<size_t>& trainLabels,
...
+  input.imbue([&count] () { return ++count; });
+
+  const auto value = Split(input, 0.2, false);
+  BOOST_REQUIRE_EQUAL(std::get<0>(value).n_cols, 8); // Train data
```
Comment (zoq) [inline src/mlpack/tests/split_data_test.cpp]:
```suggestion
  BOOST_REQUIRE_EQUAL(std::get<0>(value).n_cols, 8); // Train data.
```

Missed this before.

File: src/mlpack/tests/split_data_test.cpp [hunk 12]
```diff
@@ -94,20 +94,35 @@ void CheckDuplication(const Row<size_t>& trainLabels,
...
+  BOOST_REQUIRE_EQUAL(std::get<1>(value).n_cols, 2); // Test data
```
Comment (zoq) [inline src/mlpack/tests/split_data_test.cpp]:
```suggestion
  BOOST_REQUIRE_EQUAL(std::get<1>(value).n_cols, 2); // Test data.
```

Missed this as well.

Comment (jeffin143) [reply src/mlpack/tests/split_data_test.cpp]: Done

File: src/mlpack/methods/preprocess/preprocess_split_main.cpp [hunk 13]
```diff
@@ -73,6 +80,7 @@ PARAM_DOUBLE_IN("test_ratio", "Ratio of test set; if not set,"
     "the ratio defaults to 0.2", "r", 0.2);
 
 PARAM_INT_IN("seed", "Random seed (0 for std::time(NULL)).", "s", 0);
+PARAM_FLAG("shuffle_data", "Avoid shuffling and spliting the data.", "S");
```
Comment (favre49) [inline src/mlpack/methods/preprocess/preprocess_split_main.cpp]:
```suggestion
PARAM_FLAG("shuffle_data", "Avoid shuffling and splitting the data.", "S");
```

Also, this indicates that if `shuffle_data` is true, we avoid shuffling and splitting the data, which isn't true. I think the description should be `Shuffle and split the data`

Comment (favre49) [reply src/mlpack/methods/preprocess/preprocess_split_main.cpp]: Oops I'm sorry, I didn't read further, you've complemented it. However, the name still seems misleading to me. I would intuitively assume that `shuffle_data` means that it shuffles data if true.

File: src/mlpack/core/data/split_data.hpp [hunk 14]
```diff
@@ -61,9 +64,17 @@ void Split(const arma::Mat<T>& input,
...
+  if (shuffleData)
+  {
+    order = arma::shuffle(arma::linspace<arma::Col<size_t>>(0,
+        input.n_cols - 1, input.n_cols));
+  }
+  else
+  {
+    order = arma::linspace<arma::Col<size_t>>(0, input.n_cols - 1,
+        input.n_cols);
+  }
```
Comment (rcurtin) [inline src/mlpack/core/data/split_data.hpp]: Just one comment from me---would you mind redoing the implementation so that `order` is not used at all if `shuffleData` is `false`?  This implementation will unnecessarily create a (potentially big) vector `order` and do a bunch of unnecessary work, when we could simplify and just use a direct call to `input.cols()` and `inputLabel.subvec()`.

Comment (jeffin143) [reply src/mlpack/core/data/split_data.hpp]: Done :) ✅

File: src/mlpack/core/data/split_data.hpp [hunk 15]
```diff
@@ -52,7 +54,8 @@ void Split(const arma::Mat<T>& input,
            arma::Mat<T>& testData,
            arma::Row<U>& trainLabel,
            arma::Row<U>& testLabel,
-           const double testRatio)
+           const double testRatio,
+           bool shuffleData = true)
```
Comment (rcurtin) [inline src/mlpack/core/data/split_data.hpp:58]:
```suggestion
           const bool shuffleData = true)
```
(Just for consistency with the other parameters. :))

There are a couple places where this comment applies, but I've only highlighted the first.

File: src/mlpack/core/data/split_data.hpp [hunk 16]
```diff
@@ -61,20 +64,28 @@ void Split(const arma::Mat<T>& input,
...
-                                                      input.n_cols));
-
-  for (size_t i = 0; i != trainSize; ++i)
+  if (shuffleData)
   {
-    trainData.col(i) = input.col(order[i]);
-    trainLabel(i) = inputLabel(order[i]);
+    arma::Col<size_t> order;
+    order = arma::shuffle(arma::linspace<arma::Col<size_t>>(0,
+        input.n_cols - 1, input.n_cols));
```
Comment (rcurtin) [inline src/mlpack/core/data/split_data.hpp:71]:
```suggestion
    arma::uvec order = arma::shuffle(arma::linspace<arma::uvec>(
        0, input.n_cols - 1, input.n_cols);
```

Can we use the old implementation?  (This happens a few times in the code.) It's *basically* the same but uses one fewer line, and the `const` is a good idea for the compiler.

I changed the type to `uvec` to match the next suggestion... :)

File: src/mlpack/core/data/split_data.hpp [hunk 17]
```diff
@@ -61,20 +64,28 @@ void Split(const arma::Mat<T>& input,
...
+    for (size_t i = 0; i != trainSize; ++i)
+    {
+      trainData.col(i) = input.col(order[i]);
+      trainLabel(i) = inputLabel(order[i]);
+    }
+    for (size_t i = 0; i != testSize; ++i)
+    {
+      testData.col(i) = input.col(order[i + trainSize]);
+      testLabel(i) = inputLabel(order[i + trainSize]);
+    }
```
Comment (rcurtin) [inline src/mlpack/core/data/split_data.hpp:81]: Now that we require Armadillo 8.400.0, which has non-contiguous subviews, I think we can do what I've written below, which is really cool!

```suggestion
    trainData = input.cols(order.subvec(1, trainSize));
    trainLabel = inputLabel.subvec(order.subvec(1, trainSize));

    testData = input.cols(order.subvec(trainSize, input.n_cols - 1));
    testLabel = inputLabels.subvec(order.subvec(trainSize, input.n_cols - 1));
```

Be sure to double-check that code... I'm not 100% sure it works...

Also, I didn't suggest the change for the other one or two places where code like this occurs, so if this works, could you apply the idea there too? :)

File: src/mlpack/core/data/split_data.hpp [hunk 18]
```diff
@@ -44,6 +44,8 @@ namespace data {
  * @param trainLabel Vector to store training labels into.
  * @param testLabel Vector to store test labels into.
  * @param testRatio Percentage of dataset to use for test set (between 0 and 1).
+ * @param shuffleData If true, the sample order is shuffled; otherwise, each
+ *       sample is visited in linear order. (Default true).
```
Comment (rcurtin) [inline src/mlpack/core/data/split_data.hpp:48]:
```suggestion
 *       sample is visited in linear order. (Default true.)
```

This is super pedantic, sorry about that. :)

File: src/mlpack/methods/preprocess/preprocess_split_main.cpp [hunk 19]
```diff
@@ -73,6 +80,7 @@ PARAM_DOUBLE_IN("test_ratio", "Ratio of test set; if not set,"
     "the ratio defaults to 0.2", "r", 0.2);
 
 PARAM_INT_IN("seed", "Random seed (0 for std::time(NULL)).", "s", 0);
+PARAM_FLAG("shuffle_data", "Avoid shuffling and splitting the data.", "S");
```
Comment (rcurtin) [inline src/mlpack/methods/preprocess/preprocess_split_main.cpp:83]: If this flag indicates that we *shouldn't* shuffle, should we name it something like `no_shuffle`?

File: src/mlpack/tests/main_tests/preprocess_split_test.cpp [hunk 20]
```diff
@@ -199,4 +199,35 @@ BOOST_AUTO_TEST_CASE(PreprocessSplitUnityTestRatioTest)
       labelSize);
 }
 
+/**
+ * Check shuffle_data flag is working as expected.
+ */
+BOOST_AUTO_TEST_CASE(PreprocessSplitLabelSuffleDataTest)
```
Comment (rcurtin) [inline src/mlpack/tests/main_tests/preprocess_split_test.cpp:205]:
```suggestion
BOOST_AUTO_TEST_CASE(PreprocessSplitLabelShuffleDataTest)
```
:)

File: src/mlpack/tests/main_tests/preprocess_split_test.cpp [hunk 21]
```diff
@@ -199,4 +199,35 @@ BOOST_AUTO_TEST_CASE(PreprocessSplitUnityTestRatioTest)
...
+
+  // Now check that the output has desired dimensions.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("training").n_cols,
+      std::ceil(0.9 * inputSize));
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("test").n_cols,
+      std::floor(0.1 * inputSize));
+
+  arma::mat concat = arma::join_rows(CLI::GetParam<arma::mat>("training"),
+      CLI::GetParam<arma::mat>("test"));
+  CheckMatrices(inputData, concat);
```
Comment (rcurtin) [inline src/mlpack/tests/main_tests/preprocess_split_test.cpp:230]: Nice test, very straightforward. :+1:

File: src/mlpack/tests/split_data_test.cpp [hunk 22]
```diff
@@ -94,20 +94,35 @@ void CheckDuplication(const Row<size_t>& trainLabels,
...
   mat input(2, 10);
-  size_t count = 0; // count for putting unique sequential values
+  size_t count = 0; // Counter for unique sequential values.
   input.imbue([&count] () { return ++count; });
 
   const auto value = Split(input, 0.2);
-  BOOST_REQUIRE_EQUAL(std::get<0>(value).n_cols, 8); // train data
-  BOOST_REQUIRE_EQUAL(std::get<1>(value).n_cols, 2); // test data
+  BOOST_REQUIRE_EQUAL(std::get<0>(value).n_cols, 8); // Train data.
+  BOOST_REQUIRE_EQUAL(std::get<1>(value).n_cols, 2); // Test data.
```
Comment (rcurtin) [inline src/mlpack/tests/split_data_test.cpp:105]: Thanks, I appreciate the comment fixes here. :)

File: HISTORY.md [hunk 23]
```diff
@@ -89,7 +91,7 @@
   * Add Soft Shrink Activation Function (#2174).
 
   * Add Hinge Embedding Loss Function (#2229).
-
+
```
Comment (zoq) [inline HISTORY.md:94]:
```suggestion
 
```
My mistake.

File: src/mlpack/core/data/split_data.hpp [hunk 24]
```diff
@@ -98,29 +109,39 @@ void Split(const arma::Mat<T>& input,
...
-      arma::shuffle(arma::linspace<arma::Col<size_t>>(0, input.n_cols -1,
-                                                      input.n_cols));
-
-  for (size_t i = 0; i != trainSize; ++i)
+  if (shuffleData)
   {
-    trainData.col(i) = input.col(order[i]);
+    arma::Col<size_t> order;
+    order = arma::shuffle(arma::linspace<arma::Col<size_t>>(0,
+        input.n_cols -1, input.n_cols));
```
Comment (zoq) [inline src/mlpack/core/data/split_data.hpp:131]:
```suggestion
        input.n_cols - 1, input.n_cols));
```

File: src/mlpack/core/data/split_data.hpp (hunk 15 above)
Comment (jeffin143) [reply src/mlpack/core/data/split_data.hpp]: Done

File: src/mlpack/core/data/split_data.hpp (hunk 16 above)
Comment (jeffin143) [reply src/mlpack/core/data/split_data.hpp]: Changed

File: src/mlpack/core/data/split_data.hpp (hunk 17 above)
Comment (jeffin143) [reply src/mlpack/core/data/split_data.hpp]: There was slight issues, changed it accordingly

File: src/mlpack/tests/main_tests/preprocess_split_test.cpp (hunk 20 above)
Comment (jeffin143) [reply src/mlpack/tests/main_tests/preprocess_split_test.cpp]: Resolved

File: src/mlpack/methods/preprocess/preprocess_split_main.cpp (hunk 19 above)
Comment (jeffin143) [reply src/mlpack/methods/preprocess/preprocess_split_main.cpp]: thanks for the naming, it handled @favre49 issue on this pr also :)

File: src/mlpack/tests/main_tests/preprocess_split_test.cpp (hunk 21 above)
Comment (jeffin143) [reply src/mlpack/tests/main_tests/preprocess_split_test.cpp:230]: Thank you for this great library :)

File: src/mlpack/tests/split_data_test.cpp (hunk 22 above)
Comment (jeffin143) [reply src/mlpack/tests/split_data_test.cpp:105]: 👍

File: src/mlpack/core/data/split_data.hpp (hunk 24 above)
Comment (jeffin143) [reply src/mlpack/core/data/split_data.hpp]: changed

File: src/mlpack/methods/preprocess/preprocess_split_main.cpp (hunk 13 above)
Comment (jeffin143) [reply src/mlpack/methods/preprocess/preprocess_split_main.cpp]: Changed it `no_shuffle` , I agree but I ain't good at naming 😉

File: src/mlpack/core/data/split_data.hpp (hunk 18 above)
Comment (jeffin143) [reply src/mlpack/core/data/split_data.hpp]: Did

**Comment by zoq** [comment]
Let me know if you need help to debug the failing test.

File: src/mlpack/core/data/split_data.hpp [hunk 25]
```diff
@@ -61,20 +64,21 @@ void Split(const arma::Mat<T>& input,
...
+  if (shuffleData)
   {
-    trainData.col(i) = input.col(order[i]);
-    trainLabel(i) = inputLabel(order[i]);
+    arma::uvec order = arma::shuffle(arma::linspace<arma::uvec>(
+        0, input.n_cols - 1, input.n_cols));
+    trainData = input.cols(order.subvec(0, trainSize - 1));
+    trainLabel = inputLabel.cols(order.subvec(0, trainSize - 1));
+    testData = input.cols(order.subvec(trainSize, input.n_cols - 1));
+    testLabel = inputLabel.cols(order.subvec(trainSize, input.n_cols - 1));
```
Comment (rcurtin) [inline src/mlpack/core/data/split_data.hpp:74]: To fix the failing tests, we have to account for when `ratio` is `0` or `1`.  In these edge cases, `trainSize` will be greater than `input.n_cols - 1` (causing an error), or `trainSize` will be 0.  So, if you encase these blocks in, e.g., `if (trainSize > 0) ...` and `if (trainSize < input.n_cols)` ... then I think this should fix the issue. :+1:

Comment (jeffin143) [reply src/mlpack/core/data/split_data.hpp]: How are you so good ?

How did you come up with it ?

Comment (rcurtin) [reply src/mlpack/core/data/split_data.hpp]: Oh, I didn't do anything special. :) I checked out the branch and ran the test; when I saw it was failing, I started `gdb` and then had it break on any exceptions with `catch throw`.  Then, I ran the test, and when it threw the exception, I just looked at where the code was in `Split()`.  Quick inspection of `trainSize` and `input.n_cols` showed that `trainSize` was larger than `input.n_cols`, so then I realized what was happening. :)

Comment (jeffin143) [reply src/mlpack/core/data/split_data.hpp]: I added a test , so that next time someone changed implementation they take care of these edge cases

**Comment by jeffin143** [comment]
All Green :)

**Comment by rcurtin** [comment]
Thanks @jeffin143! :+1:
